packages/react-server-dom-vite/src/ReactFlightClientConfigViteBundler.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {
+  Thenable,
+  FulfilledThenable,
+  RejectedThenable,
+} from 'shared/ReactTypes';
+
+export type SSRManifest = string; // Module root path
+
+export type ServerManifest = string; // Module root path
+
+export type ServerReferenceId =
+  | [
+      string, // module path
+      string, // export name
+    ]
+  | string;
+
+export opaque type ClientReferenceMetadata = [
+  string, // module path
+  string, // export name
+];
+
+// eslint-disable-next-line no-unused-vars
+export opaque type ClientReference<T> = {
+  specifier: string,
+  name: string,
+};
+
+export function resolveClientReference<T>(
+  bundlerConfig: SSRManifest,
+  metadata: ClientReferenceMetadata,
+): ClientReference<T> {
+  return {
+    specifier: metadata[0],
+    name: metadata[1],
+  };
+}
+
+export function resolveServerReference<T>(
+  config: ServerManifest,
+  id: ServerReferenceId,
+): ClientReference<T> {
+  if (typeof id === 'string') {
+    const [specifier, name] = id.split(',');
+    return {
+      specifier,
+      name,
+    };
+  }
+  return {
+    specifier: id[0],
+    name: id[1],
+  };
+}
+
+export function preloadModule<T>(
+  metadata: ClientReference<T>,
+): null | Thenable<any> {
+  const existingPromise = __vite_module_cache__.get(metadata.specifier);
+  if (existingPromise) {
+    if (existingPromise.status === 'fulfilled') {
+      return null;
+    }
+    return existingPromise;
+  } else {
+    // $FlowFixMe[unsupported-syntax]
+    const modulePromise: Thenable<T> = __vite_require__(metadata.specifier);
+    modulePromise.then(
+      value => {
+        const fulfilledThenable: FulfilledThenable<mixed> =
+          (modulePromise: any);
+        fulfilledThenable.status = 'fulfilled';
+        fulfilledThenable.value = value;
+      },
+      reason => {
+        const rejectedThenable: RejectedThenable<mixed> = (modulePromise: any);
+        rejectedThenable.status = 'rejected';
+        rejectedThenable.reason = reason;
+      },
+    );
+    __vite_module_cache__.set(metadata.specifier, modulePromise);
+    return modulePromise;
+  }
+}
+
+export function requireModule<T>(metadata: ClientReference<T>): T {
+  let moduleExports;
+  // We assume that preloadModule has been called before, which
+  // should have added something to the module cache.
+  const promise: any = __vite_module_cache__.get(metadata.specifier);
```

Comment:
Note that the Webpack implementation only uses the preinitialization for “async modules” which is not most modules.

For most modules the module is “required” aka initialized synchronously at this point - lazily.

By moving all modules to be eagerly initialized in the preload, this will have much worse performance and scheduling capability than the webpack one.

I’m not sure if that’s inherent in the output of Vite. For dev maybe it doesn’t matter much but it would be relevant for production.

Is there anyway to do sync module evaluation in Vite?

Reply:
I don't think during development that's an option. I think it'll have to be async during development. Vite doesn't use node loaders or anything like that. Instead we use `vite.ssrLoadModule` that resolves, loads and transforms a module using the vite plugins and delivers that. But during production, if we can build something that can be consumed via `require`, we can do sync evaluation. Right now its not setup like that. We are building out ESM bundles for node too and so we need dynamic import (async) to load those.

Reply:
Also what makes a module an async module in webpack's case?

Reply:
https://webpack.js.org/blog/2020-10-10-webpack-5-release/#async-modules

Reply:
resolved in last change. actually compile to CJS for the node servers and require the modules syncronously